### PR TITLE
Add errors-only output mode

### PR DESCRIPTION
**Pull request by dobesv**
For users that are not that interested in all the output related cache hits/misses and the cached output of prior builds or even the new output from a new build step that's just print a count of files or something boring like that, this trims the output down to just the task that failed, if any.

Partial solution for https://github.com/vercel/turbo/issues/993

### CODE REVIEW

File: cli/internal/runcache/runcache.go [hunk 1]
```diff
@@ -198,17 +198,32 @@ func (tc TaskCache) RestoreOutputs(ctx context.Context, prefixedUI *cli.Prefixed
...
 
+// ReplayLogFile writes out the stored logfile to the terminal
+func (tc TaskCache) ReplayLogFile(prefixedUI *cli.PrefixedUi, progressLogger hclog.Logger) {
+	if tc.LogFileName.FileExists() {
+		tc.rc.logReplayer(progressLogger, prefixedUI, tc.LogFileName)
+	}
+}
+
+// OnError
+// If output mode is set to log output only after error, we'll have buffered output
```
Comment (mehulkar) [inline cli/internal/runcache/runcache.go:219]:
```suggestion
// If output-mode is set to "errors-only", we'll have buffered output
```

File: docs/components/output-mode-table.mdx [hunk 2]
```diff
@@ -1,6 +1,7 @@
...
-| hash-only | Show only the hashes of the tasks        |
-| new-only  | Only show output from cache misses       |
-| none      | Hides all task output                    |
+| option      | description                              |
+| ----------- | ---------------------------------------- |
+| full        | This is the default. Displays all output |
+| hash-only   | Show only the hashes of the tasks        |
+| new-only    | Only show output from cache misses       |
+| errors-only | Only show output from task failures      |
+| none        | Hides all task output                    |
```
Comment (mehulkar) [inline docs/components/output-mode-table.mdx:7]: Can you move docs to a separate PR? Since the feature won't be out until the next release, we'll hold onto the docs PR until then.

**Review by mehulkar — COMMENTED** [review]
This looks good to me. An integration test under `cli/integration-tests` that checks log output would be a good addition here.

**Comment by dobesv** [comment]
Thanks for the review!

I rolled back the docs changes and added a couple of tests.  I'm not too familiar with the testing system and philosophy here so do let me know how to do the tests properly if I didn't do them right.

File: cli/integration_tests/monorepo_one_script_error/run.t [hunk 3]
```diff
@@ -52,3 +52,27 @@ Make sure it isn't cached
...
+  my-app:error: npm ERR!   in workspace: my-app 
+  my-app:error: npm ERR!   at location: .*/run.t/apps/my-app  (re)
+  command \(.*/run.t/apps/my-app\) npm run error exited \(1\) (re)
+  
+   Tasks:    0 successful, 1 total
+  Cached:    0 cached, 1 total
+    Time:\s*[\.0-9]+m?s  (re)
+  
+   ERROR  run failed: command  exited (1)
+  [1]
```
Comment (mehulkar) [inline cli/integration_tests/monorepo_one_script_error/run.t:78]: hmm, I think we should also have a test case where there are two scripts that run, one that errors and one that doesn't. You could add to the `monorepo` fixture in this same integration test or add a totally new `cli/integration_test/monorepo_errors_only_output`, I don't have a strong preference!

File: cli/integration_tests/single_package_deps/run.t [hunk 4]
```diff
@@ -39,4 +39,33 @@ Run a second time, verify caching works because there is a config
...
+  
+Run with --output-logs=none
+  $ ${TURBO} run test --single-package --output-logs=none
+  \xe2\x80\xa2 Running test (esc)
+  \xe2\x80\xa2 Remote caching disabled (esc)
+  
+   Tasks:    2 successful, 2 total
+  Cached:    2 cached, 2 total
+    Time:\s*[\.0-9]+m?s >>> FULL TURBO (re)
+
```
Comment (mehulkar) [inline cli/integration_tests/single_package_deps/run.t:71]: This is great!

**Review by mehulkar — COMMENTED** [review]
Almost there, thanks for the quick turnaround and working with me on this!

File: cli/integration_tests/monorepo_one_script_error/run.t (hunk 3 above)
Comment (dobesv) [reply cli/integration_tests/monorepo_one_script_error/run.t:78]: OK, I added another task to this folder, seemed the simplest.

File: cli/integration_tests/monorepo_one_script_error/run.t [hunk 5]
```diff
@@ -33,7 +33,7 @@ Make sure it isn't cached
   \xe2\x80\xa2 Packages in scope: my-app (esc)
   \xe2\x80\xa2 Running error in 1 packages (esc)
   \xe2\x80\xa2 Remote caching disabled (esc)
-  my-app:error: cache miss, executing aed5e5ad070dbe8e
+  my-app:error: cache miss, executing [0-9a-f]+ (re)
```
Comment (mehulkar) [inline cli/integration_tests/monorepo_one_script_error/run.t:36]: This is clever, but I think we actually care to fail tests when the hash changes. Could you make it the real thing?

File: cli/integration_tests/monorepo_one_script_error/run.t [hunk 6]
```diff
@@ -52,3 +52,27 @@ Make sure it isn't cached
   
    ERROR  run failed: command  exited (1)
   [1]
+
+Running with --output-mode=errors-only gives error output only
+  $ ${TURBO} --output-logs=errors-only error okay
```
Comment (mehulkar) [inline cli/integration_tests/monorepo_one_script_error/run.t:57]: nice 👍🏾

**Review by mehulkar — COMMENTED** [review]
Thanks for the quick update and apologies for not getting this merged yesterday. One more comment about the test case and merge conflict needs resolved. 🤞🏾 we can get this in today! (I'll watch for updates here)

File: cli/internal/runcache/runcache.go [hunk 7]
```diff
@@ -198,17 +198,32 @@ func (tc TaskCache) RestoreOutputs(ctx context.Context, prefixedUI *cli.Prefixed
...
+func (tc TaskCache) ReplayLogFile(prefixedUI *cli.PrefixedUi, progressLogger hclog.Logger) {
+	if tc.LogFileName.FileExists() {
+		tc.rc.logReplayer(progressLogger, prefixedUI, tc.LogFileName)
+	}
+}
+
+// OnError
+// If output-mode is set to "errors-only", we'll have buffered output
+// to disk without logging it.  Now we can log that here.
+func (tc TaskCache) OnError(terminal *cli.PrefixedUi, logger hclog.Logger) {
```
Comment (mehulkar) [inline cli/internal/runcache/runcache.go:221]: This is annoying, but lint requires the code comment to be a whole line starting with `OnError`. You can still break it up into multiple lines, but I guess specifically what you have here is not acceptable 🤦🏾 ...

**Comment by mehulkar** [comment]
Merged in #2672. Thanks @dobesv !

**Comment by dobesv** [comment]
> Merged in https://github.com/vercel/turbo/pull/2672

Oh, uh thanks!  I hadn't fixed the linter error, thanks for cleaning that up for me
